node/libs/roles/src/validator/messages/consensus.rs

```diff
@@ -70,6 +70,29 @@ impl Default for Fork {
     }
 }
 
+/// The mode used for selecting leader for a given view.
+#[derive(Default, Clone, Debug, Eq, PartialEq)]
+pub enum LeaderSelectionMode {
+    /// Select in a round-robin fashion, based on validators' index within the set.
+    #[default]
+    RoundRobin,
+
+    /// Select based on a sticky assignment to a specific validator.
+    Sticky(validator::PublicKey),
+
+    /// Select pseudo-randomly, based on validators' weights.
+    Weighted,
+}
+
+/// Calculates the pseudo-random eligibility of a leader based on the input and total weight.
+pub(crate) fn leader_weighted_eligibility(input: u64, total_weight: u64) -> u64 {
+    let input_bytes = input.to_be_bytes();
+    let hash = Keccak256::new(&input_bytes);
+    let hash_64 = &hash.as_bytes()[0..8];
+    let hash_num = u64::from_be_bytes(hash_64.try_into().unwrap());
+    hash_num % total_weight
```

Comment:
this doesn't generate a uniform distribution for large weights

Reply:
For `total_weight` which is `u64::MAX`, it will return the value of the hash first 8 bytes. Is that a problem?
Other option is to convert both values to big uint, perform the modulo, and then convert back to a single u64, knowing it should be less then that.

Reply:
I mean that if you have 2 validators each with weight u64::MAX/3, then the first validator will get 2/3 of the blocks.

Reply:
i.e. IMO we should make the distribution error significantly smaller for every possible inputs.

Reply:
Thanks for the clarification. The solution I suggested was implemented here: https://github.com/matter-labs/era-consensus/pull/96/commits/feabe397ef86e49db3ca249bc2d957080fd446e3
I assume that using the hash output's entire 256 bits to reduce non-uniformity is sufficient. LMK if I'm still missing something.

Reply:
LG, as far as I'm concerned u128 would also be good enough, but u256 is good as well.